environment/environment.go

```diff
@@ -0,0 +1,71 @@
+package environment
+
+import (
+	"fmt"
+
+	"github.com/sourcegraph/conc"
+)
+
+type deferFn struct {
+	msg string
+	fn  func() error
+}
+
+// Env controls goroutines and deferred functions. Think of it as the "main thread."
+// It is not goroutine-safe.
+type Env struct {
+	deferFns []*deferFn
+	wg       *conc.WaitGroup
+}
```

Comment:
Possibly, though I think `Close` being called twice is a sign of poorly written code. Because the usage of `Env` is pretty straightforward, I don't think we need to be as defensive here, especially since the warning about `Close` is documented.

I'll leave it as-is for now and we can add something like this later if it becomes an issue.